app/models/king.rb

```diff
@@ -1,7 +1,68 @@
+# rubocop:disable Metrics/AbcSize
 class King < Piece
   def move_legal?(to_row, to_col)
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
     row_diff <= 1 && col_diff <= 1
+    # add castling moves to the 'whitelist', then verify the validity of the castle
```

Comment:
Self checks are handled after #valid_move. I've added a #finalize_move! that first calls #self_check?, then calls #move_to!.

The alternative of calling #self_check? in #valid_move? broke several tests, and causes a call loop when two kings are within checking distance of each other.